src/lib/mlang.js

```diff
@@ -73,72 +75,182 @@ foam.CLASS({
 });
 
 
+/**
+ * Base class for all mLang queries.
+ *
+ * Contains many default implementations of methods.
+ */
 foam.CLASS({
-  package: 'foam.mlang',
-  name: 'Expr'
+  package: 'foam.mlang.predicate',
+  name: 'Expr',
+  abstract: true,
+  // PORT: There should be several interfaces here eventually, eg.
+  // Functor/Projectable/whatever we call f()'s interface.
+  // implements: ['ExprProtocol'],
+
+  methods: [
+    /**
+     * Abstract projection method. Given an object, extracts the value of a
+     * particular property from it.
+     */
+    {
+      name: 'f',
+      args: [
+        {
+          name: 'obj',
+        },
+      ],
+    },
+    /**
+     * Converts this expression into MQL syntax, as used in eg. Gmail searches.
+     */
+    function toMQL() {
+      return this.cls_.label;
+    },
+    /** Converts this expression into SQL, suitable for use in WHERE clauses. */
+    function toSQL() {
+      return this.cls_.label;
+    },
+    /** Converts this expression to a human-readable string for debugging. */
+    function toString() {
+      return this.toMQL();
+    },
+    /** Recursively adds all inputs of an expression to an array. */
+    function collectInputs(terms) {
+      terms.push(this);
+    },
+    /**
+     * Simplifies an expression by eliminating unnecessary clauses, and
+     * combining others. Can sometimes reduce whole (sub)expressions to TRUE or
+     * FALSE.
+     */
+    function partialEval() {
+      return this;
+    },
+
+    // PORT: pipe() is kind of the wrong name, this is a different meaning and
+    // signature from DAO.pipe(), which could be very confusing. chain()?
+    // wrap()?
+    /**
+     * Returns a Sink which applies this expression to each value put or
+     * removed, calling the provided sink only for those that match this
+     * expression.
+     */
+    /*
+    function pipe(sink) {
```

Comment:
I would rather delete this as well and put this as a TODO in the bottom of the file, like we did for mm.js/dao.js